reactive/kotlinx-coroutines-rx2/src/RxScheduler.kt

```diff
@@ -4,16 +4,118 @@
 
 package kotlinx.coroutines.rx2
 
-import io.reactivex.Scheduler
+import io.reactivex.*
+import io.reactivex.disposables.*
+import io.reactivex.plugins.*
 import kotlinx.coroutines.*
-import java.util.concurrent.TimeUnit
-import kotlin.coroutines.CoroutineContext
+import kotlinx.coroutines.channels.*
+import java.util.concurrent.*
+import kotlin.coroutines.*
 
 /**
  * Converts an instance of [Scheduler] to an implementation of [CoroutineDispatcher]
  * and provides native support of [delay] and [withTimeout].
  */
-public fun Scheduler.asCoroutineDispatcher(): SchedulerCoroutineDispatcher = SchedulerCoroutineDispatcher(this)
+public fun Scheduler.asCoroutineDispatcher(): CoroutineDispatcher =
+    if (this is DispatcherScheduler) {
+        dispatcher
+    } else {
+        SchedulerCoroutineDispatcher(this)
+    }
+
+/**
+ * Converts an instance of [CoroutineDispatcher] to an implementation of [Scheduler].
+ */
+public fun CoroutineDispatcher.asScheduler(): Scheduler =
+    if (this is SchedulerCoroutineDispatcher) {
+        scheduler
+    } else {
+        DispatcherScheduler(this)
+    }
+
+private class DispatcherScheduler(internal val dispatcher: CoroutineDispatcher) : Scheduler() {
+
+    val parentJob = SupervisorJob()
+    private val parentScope = CoroutineScope(parentJob)
+
+    override fun scheduleDirect(run: Runnable): Disposable {
+        val decoratedRun = RxJavaPlugins.onSchedule(run)
+        val worker = createWorker() as DispatcherWorker
+        return worker.apply {
+            worker.schedule(decoratedRun)
+        }
+    }
+
+    override fun scheduleDirect(run: Runnable, delay: Long, unit: TimeUnit): Disposable {
+        val decoratedRun = RxJavaPlugins.onSchedule(run)
+        val worker = createWorker() as DispatcherWorker
+        return worker.apply {
+            schedule(decoratedRun, delay, unit)
+        }
+    }
+
+    override fun createWorker(): Worker =
+        DispatcherWorker(dispatcher, parentJob)
+
+    override fun shutdown() {
+        parentScope.cancel()
+    }
+
+    private class DispatcherWorker(private val dispatcher: CoroutineDispatcher, parentJob: Job) : Worker() {
+
+        private val workerScope = CoroutineScope(SupervisorJob(parentJob) + dispatcher)
+        private val blockChannel = Channel<Runnable>()
+        private var queueProcessingJob: Job? = null
+
+        override fun isDisposed(): Boolean = !workerScope.isActive
+
+        override fun schedule(block: Runnable): Disposable {
+            startProcessingQueue()
+            if (workerScope.isActive) {
+                workerScope.launch(dispatcher) {
```

Comment:
I think this could violate the Worker contract. If the underlying dispatcher uses a thread pool, and a bunch of tasks are scheduled on the worker at the same time, these launched coroutines could get distributed over multiple threads and the order in which the tasks are added to the queue and executed would be non-deterministic.

One solution would be to make the channel unlimited capacity and just call `blockChannel.offer` here, although you'd need to then explicitly associate some disposable with the Runnable.

Reply:
Good point. I actually found what I think is a good solution to this by offering a lazy ```Job```. Let me know what you think (about to push in a few)

Reply:
@zach-klippenstein done. let me know what you think of new implementation.